File: Python/problem4.py
def reverse_string(string: str) -> str:
    return string[::-1]


def is_palindrome(number: str) -> bool:
    number_reverse: str = reverse_string(number)
    return number == number_reverse


def solve(start: int, stop: int) -> int:
    max_product: int = 0
    for i in range(start, stop):
        for j in range(start, stop):
            product: int = i * j
            if is_palindrome(str(product)) and product > max_product:
                max_product = product
    return max_product


if __name__ == "__main__":
    print(solve(100, 1000))

File: C++/Problem4.c
#include <iostream>
using namespace std;

int number_of_digits(int number)
{
  int len = 0;
  do
  {
    number /= 10;
    len++;
  } while (number);
  return len;
}

bool isPalindrome(int number)
{
  int length = number_of_digits(number);
  int list_number[length];
  int i = 0;
  while (i < length)
  {
    list_number[i] = (number % 10);
    number = number / 10;
    i++;
  }
  for (i = 0; i < length; i++)
  {
    if (list_number[i] != list_number[length-1-i])
    {
      return false;
    }
  }
  return true;
}

int main()
{
  int max_palindrome = 0;
  for (int i = 100; i < 1000; i++)
  {
    for (int j = 100; j < 1000; j++)
    {
      if (isPalindrome(i*j))
      {
        if (i*j > max_palindrome) max_palindrome = i*j;
      }
    }
  }
  cout << max_palindrome << "\n";
  
}


